src/bin/pg_autoctl/primary_standby.c

```diff
@@ -1456,3 +1456,91 @@ standby_cleanup_as_primary(LocalPostgresServer *postgres)
 
 	return true;
 }
+
+
+/*
+ * standby_check_timeline_with_upstream returns true when the current timeline
+ * on the local node (a standby) is the same as the timeline fetched on the
+ * upstream node setup in its replicationSource.
+ */
+bool
+standby_check_timeline_with_upstream(LocalPostgresServer *postgres)
+{
+	ReplicationSource *replicationSource = &(postgres->replicationSource);
+	NodeAddress *primaryNode = &(replicationSource->primaryNode);
+
+	/* pg_control_checkpoint() returns the local timeline, update it now */
+	if (!pgsql_checkpoint(&(postgres->sqlClient)))
```

Comment:
Creating a checkpoint can be quite heavy. If we only do this to get the current timeline id, maybe instead we should use some other function, such as `pg_walfile_name` or `pg_control_data`: https://www.postgresql.org/message-id/3FB63E54-403A-440E-88CB-CFA6FEF3E0F7%40icloud.com

Reply:
We can grab `pg_stat_wal_receiver.received_tli`, I believe that having received it locally is what matters in our case. If we have to restart (crash recovery or normal startup), we will be able to catch-up with the new tli, and more importantly in case of a promotion, again, we can replay the WAL files and switch to the new TLI then.

Note that using `pg_walfile_name` requires knowing which WAL file is the most recently received, and should give us the same information as `pg_stat_wal_receiver` I believe, and `pg_control_data` needs a checkpoint too and shows the same value as the `pg_control_checkpoint` view I was using before.